app/src/main/java/tech/jpco/qen/viewModel/QenViewModel.kt

```diff
@@ -0,0 +1,128 @@
+package tech.jpco.qen.viewModel
+
+import android.util.Log
+import androidx.annotation.VisibleForTesting
+import androidx.lifecycle.ViewModel
+import io.reactivex.Observable
+import io.reactivex.disposables.CompositeDisposable
+import io.reactivex.subjects.BehaviorSubject
+import io.reactivex.subjects.PublishSubject
+import tech.jpco.qen.model.PagesRepository
+
+val Any.TAG
+    get() = this::class.simpleName ?: "Anon"
+
+fun Any.iLogger(output: String, obj: Any? = Unit) {
+    val name = Thread.currentThread().name
+    val objS = if (obj == Unit) "" else ": $obj"
+    Log.d(TAG, "$output$objS on ${name.substring(0, 1).toUpperCase()}${name.substring(1)}")
+}
+
+
+class QenViewModel : ViewModel() {
+    private val cd = CompositeDisposable()
+
+    //(Parallel) Dual Subject architecture per https://speakerdeck.com/oldergod/model-view-intent-for-android
+    /* In theory, it keeps an "alive" data stream from UI interaction to UI output, allowing an event in flight to
+     * be resubscribed by a recreated view before it has finished its journey.
+     */
+    //TODO test the above description!
+    private val touchesIn = PublishSubject.create<DrawPoint>()
+    private val metaIn = PublishSubject.create<MetaEvent>()
+    private val mTouchesOut = PublishSubject.create<DrawPoint>()
+    private val mMetaOut = BehaviorSubject.create<SelectedPage>()
+
+    val touchesOut: Observable<DrawPoint> = mTouchesOut
+    val metaOut: Observable<SelectedPage> = mMetaOut
+
+
+    fun supply(repo: PagesRepository, touches: Observable<DrawPoint>, events: Observable<MetaEvent>) {
```

Comment:
Rather than have this `supply` function and the `metaProcessor` function bridged together by `Subjects`, if somehow this function returned the touches and meta events that needed to be processed, then the code in here would be simpler. The bridging using the `Subjects` could be eliminated.

You could have a single function `createViewEvents(repo: PagesRepository, touches: Observable<DrawPoint>, events: Observable<MetaEvent>)`. It's return type would be of type `Observable<UiEvent>`, where `UiEvent` is a `sealed class` containing either a `DrawPoint` or a `SelectedPage`. Then the consuming code in `MainActivity.setupStreams` would switch on the type and handle each stream appropriately.

Example `UiEvent` implementation:
```kotlin
sealed class UiEvent {
  data class Draw(val drawPointL DrawPoint): UiEvent()
  data class Navigation(val selectedPage: SelectedPage): UiEvent()
}
```

Reply:
This gets to the point of the KDoc comment on lines 16 and 17. The subjects are, at least in theory, so that the View layer has something to attach to and detach from as needed for lifecycle events. Having two separate streams each for incoming and outgoing traffic was an intentional design decision, because the two don't directly interact and the DrawPoint stream emits at 20Hz (60 if I didn't rate-limit!) while the MetaEvent stream emits only occasionally...

Reply:
Okay, so looking at it further, I understand the need for the touches in and out being separate due to storing data in the repository being part of the regular flow. It's a bit confusing, as subjects are usually used for bridging imperative code to reactive code rather than hooking different reactive components together. Even though they don't really look like they are hooked together, the subjects are just re-broadcasting those events. If all the events aren't needed, throttling the incoming events like you're already doing is all you need to do. Strictly speaking, the subjects aren't needed for the drawing behavior, since the view draws perfectly fine (obviously the state saving doesn't work) if the `MainActivity` just directly subscribes the `QenPage.drawSegment` to the output of the `touchStream`.